### DISCUSSION TITLE
V18 PrimeNG styles in shadowDOM encapsulation components

### DISCUSSION THREAD

**Discussion by LMatass** [primefaces/community#2949] [PrimeNG] [answered]
Hello,

I'm using primeng in a angular webcomponent which obviusly its using shadowDOM as view encapsulation. Currently primeng its creating the style elements in the main document body, is there the possibility to change the document where the styles are created? and introduce them instead on the shadowRoot element?

Currently I have a workaround which is based on:
A dom observer that listens for the style elements with `data-primeng-style-id` attribute. Whenever one of this elements its added, I copy the element into my shadowRoot element and remove it from the document.

Is there any plan to make flexible where do we want style elements to be created? if not, is there any other workaround?

**Comment by Khanaash** [comment]
Hi,

Any updates on this thread ? I'm experiencing microfrontends with webcomponent applications (under shadowdom) and i am facing the same issues. With prime 17, we can inject the prime css by including it in the angular webcomponent css but with v18, it is automatically loaded into the doc root.

Any recommandations or workaround would be appreciated. Thank you.

**Comment by LMatass** [comment]
Right now the only workaround I found is using an iframe but these breaks the benefits of webcomponents/microfrontends... Would be great if they would offer use the possibility to reference the doc where to insert the style elements...

**Comment by shwetha-anand** [comment]
A bit atrocious that there isn't an option to inject styles to shadow root as opposed to document head . Its like we're forced to either stick with PrimeNg 17 or move out of it.

**Comment by gabriel-werzner** [comment]
any news on this issue?

**Comment by cagataycivici** [comment]
I think @mertsincan worked on this recently, Mert can you comment?

**Comment by mertsincan** [comment]
Yes, I worked on this before. I put the generated CSS variables into the **_:host_** selector too, so they reach the shadow host. But like you said, I'm not sure this is the right way to inject the styles into the shadow DOM. If the same component is used more than once, the styles get added to every shadow DOM again.

Can you try your cases with v22? If it still doesn't work, could you please attach a stackblitz link or sample project so I can take a look?

**Reply by gabriel-werzner** [reply]
That's a lot for replying, we are indeed still on v21, because of my timezone, its late at night,  I will upgrade the version to 22 tomorrow. If it doesn't work, I will create a contained environment where the issues can be replicated, if it works on v22 I will let you know. Again thanks a lot, your input is greatly appreciated.

**Reply by mertsincan** [reply]
Sure, I'll wait for your update. Thanks for testing it!

**Reply by gabriel-werzner** [reply]
Hello, I have upgraded our project to primeng v22 and the problem still persists, as you kindly suggested, I have uploaded an example to stackblitz. The dropdown shows an exception on the console when clicked, and the style doesn't translate into the component. 
https://stackblitz.com/edit/stackblitz-starters-mqznxwnd

For the styles issue, we are applying the following patch on the main.ts as a workaround for now, the js dropdown problem still persists.
https://stackblitz.com/edit/stackblitz-starters-mqk1anj3

**Reply by mertsincan** [reply]
Thanks for the examples. I’ll take a look.

**Reply by mertsincan** [reply, marked as answer]
Hi,

Sorry for the late reply, I wanted to test a few things first.

This will be fixed in 22.1.0. PrimeNG now applies its styles inside shadow roots, so you won't need the observer workaround anymore. The overlay crash is gone too.

Thanks for the report and the test project, it helped a lot.
